examples/thermostat/thermostat-common/src/atomic-write-manager.cpp

```diff
@@ -0,0 +1,576 @@
+    auto sessionHandle        = commandObj->GetExchangeContext()->GetSessionHandle();
+
+    if (sessionHandle->IsSecureSession())
+    {
+        sourceNodeId = sessionHandle->AsSecureSession()->GetPeer();
+    }
+    else if (sessionHandle->IsGroupSession())
+    {
+        sourceNodeId = sessionHandle->AsIncomingGroupSession()->GetPeer();
+    }
+    return sourceNodeId;
+}
+
+/**
+ * @brief Schedules a timer for the given timeout in milliseconds.
+ *
+ * @param[in] endpoint The endpoint to use.
+ * @param[in] timeout The timeout in milliseconds.
+ */
+void ThermostatAtomicWriteManager::ScheduleTimer(EndpointId endpoint, System::Clock::Milliseconds16 timeout)
+{
+    DeviceLayer::SystemLayer().StartTimer(timeout, TimerExpiredCallback,
+                                          reinterpret_cast<void *>(static_cast<uintptr_t>(endpoint)));
+}
+
+/**
+ * @brief Clears the currently scheduled timer.
+ *
+ * @param[in] endpoint The endpoint to use.
+ */
+void ThermostatAtomicWriteManager::ClearTimer(EndpointId endpoint)
+{
+    DeviceLayer::SystemLayer().CancelTimer(TimerExpiredCallback, reinterpret_cast<void *>(static_cast<uintptr_t>(endpoint)));
+}
+
+void ThermostatAtomicWriteManager::OnTimerExpired(EndpointId endpoint)
+{
+    for (size_t i = 0; i < ArraySize(mAtomicWriteSessions); ++i)
+    {
+
+        auto atomicWriteState = mAtomicWriteSessions[i];
+        if (atomicWriteState.state == AtomicWriteState::Open && atomicWriteState.endpointId == endpoint)
+        {
+            if (mDelegate != nullptr)
+            {
+                mDelegate->OnRollbackWrite(endpoint, Presets::Id);
+                mDelegate->OnRollbackWrite(endpoint, Schedules::Id);
+            }
+            ResetWrite(endpoint);
+        }
+    }
+}
+
+void ThermostatAtomicWriteManager::SetWriteState(EndpointId endpoint, ScopedNodeId originatorNodeId, AtomicWriteState state)
+{
+    uint16_t ep =
+        emberAfGetClusterServerEndpointIndex(endpoint, Thermostat::Id, MATTER_DM_THERMOSTAT_CLUSTER_SERVER_ENDPOINT_COUNT);
+
+    if (ep < ArraySize(mAtomicWriteSessions))
+    {
+        mAtomicWriteSessions[ep].state      = state;
+        mAtomicWriteSessions[ep].endpointId = endpoint;
+        mAtomicWriteSessions[ep].nodeId     = originatorNodeId;
+    }
+}
+
+bool ThermostatAtomicWriteManager::InWrite(const std::optional<AttributeId> attributeId, EndpointId endpoint)
+{
+    // These are the only two atomic attributes
+    if (attributeId.has_value() && attributeId.value() != Presets::Id && attributeId.value() != Schedules::Id)
+    {
+        return false;
+    }
+
+    uint16_t ep =
+        emberAfGetClusterServerEndpointIndex(endpoint, Thermostat::Id, MATTER_DM_THERMOSTAT_CLUSTER_SERVER_ENDPOINT_COUNT);
+
+    if (ep < ArraySize(mAtomicWriteSessions))
+    {
+        return mAtomicWriteSessions[ep].state == AtomicWriteState::Open;
+    }
+    return false;
+}
+
+bool ThermostatAtomicWriteManager::InWrite(const std::optional<AttributeId> attributeId,
```

Comment:
Should this logic be in the _application_ as here, or in the _cluster implementation_?

Seems to me that all the logic involving subject descriptors and whatnot is not application-specific and should be handled by the cluster implementation, because otherwise we are forcing all applications to re-implement that logic.